### PR TITLE
Wayland Layer Shell

### PR DESCRIPTION
**Pull request by gfgit**
This was extracted from #2043

### CODE REVIEW

File: panel/lxqtpanel.cpp
Comment (tsujan) [inline panel/lxqtpanel.cpp:242]: A nullity check for `windowHandle()`?

File: panel/lxqtpanel.cpp
Comment (tsujan) [inline panel/lxqtpanel.cpp:861]: Nullity check for `mLayerWindow`, as is done in other places?

File: panel/lxqtpanel.cpp
Comment (gfgit) [reply panel/lxqtpanel.cpp:239]: Hi. Good question...
I've found `create()` in [dprgadget example](https://github.com/qt/qtbase/blob/ae361f00b3f4827e62a19cc451227c9892b51f2e/tests/manual/highdpi/dprgadget/main.cpp#L197), so I've used that.
`winId()` call seems to differ only in forcing native widgets if requested.

About show/hide this is not a problem. `QWindow` is hidden but not deleted so no need to manually re-attach the `LayerShellQt` integration. It will automatically apply to the new Wayland surface.

File: panel/lxqtpanel.cpp
Comment (gfgit) [reply panel/lxqtpanel.cpp:242]: I don't think `create()` will ever fail in this scenario but it doesn't hurt anyway.

File: panel/lxqtpanel.cpp
Comment (gfgit) [reply panel/lxqtpanel.cpp:868]: Done.

File: panel/lxqtpanel.cpp
Comment (tsujan) [reply panel/lxqtpanel.cpp:239]: > About show/hide this is not a problem. QWindow is hidden but not deleted so no need to

That sounds logical, but I remember that we needed to do it before showing the widget in lxqt-runner and lxqt-notificationd. However, the reason may be in the differences between the apps; I also did in inside the c-tor in pcmanfm-qt's desktop, like you did here.

As for `create()`, looking at its code in Qt source, I thought it might be too much here.

File: panel/lxqtpanel.cpp
Comment (tsujan) [reply panel/lxqtpanel.cpp]: You're right. My request might be a little pedantic with `create()`. I thought more about `winId()`, which IMO is less "aggressive" (although I can't prove that). Anyway, nullity checks are always good :)

File: panel/lxqtpanel.cpp
Comment (gfgit) [reply panel/lxqtpanel.cpp:840]: Done.

File: panel/lxqtpanel.cpp
Comment (gfgit) [reply panel/lxqtpanel.cpp:239]: Current code works as is but both `winId()` and `create()` are ok for me.

Comment (tsujan) [reply panel/lxqtpanel.cpp:239]: > both winId() and create() are ok for me

For me too. Let's keep `create()` here to see what will happen. It was OK in my tests.

**Comment by gfgit** [comment]
Ok so remaining issue should be:
1. Panel auto-hide: here on KWin it does not let me click on bottom parts of apps (like statusbar) because panel fires on top and mouse click ends up on the panel
2. Percent lengths are glitchy here

**Comment by tsujan** [comment]
> Ok so remaining issue should be:...

Do you intend to fix them here, or do you prefer to do it in another PR, after this one is merged? I have no problem with either.

**Comment by gfgit** [comment]
Well I've already tried investigating but with no luck so far so maybe merging could be better so more people test the code.

**Comment by tsujan** [comment]
OK. I'll merge it tonight, after taking another look. Thanks!

File: panel/lxqtpanel.cpp
Comment (tsujan) [reply panel/lxqtpanel.cpp:840]: I'm afraid the warning still exists. It's about the above line (840), not the place where you edited the code.

Comment (gfgit) [reply panel/lxqtpanel.cpp]: Ops. Should be correct now.

Comment (tsujan) [reply panel/lxqtpanel.cpp]: Yes, now it fixed :)

**Review by tsujan — APPROVED** [review]
GTM.

**Pull request merged**
